test/mpi/dtpools/include/dtpools_internal.h

```diff
@@ -0,0 +1,130 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil ; -*- */
+/*
+ *
+ *  (C) 2018 by Argonne National Laboratory.
+ *      See COPYRIGHT in top-level directory.
+ */
+
+#ifndef DTPOOLS_INTERNAL_H_INCLUDED
+#define DTPOOLS_INTERNAL_H_INCLUDED
+
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include "mpi.h"
+#include "dtpools.h"
+
+#define ERR_STRING_MAX_LEN (512)
+
+typedef enum {
+    DTPI_OBJ_TYPE__BASIC,
+    DTPI_OBJ_TYPE__CONTIG,
+    DTPI_OBJ_TYPE__VECTOR,
+    DTPI_OBJ_TYPE__BLOCK_INDEXED,
+    DTPI_OBJ_TYPE__INDEXED,
+    DTPI_OBJ_TYPE__STRUCT,
+    DTPI_OBJ_TYPE__NUM
+} DTPI_obj_type_e;
+
+/* basic derived datatype structures */
+enum {
+    DTPI_OBJ_LAYOUT_SIMPLE__BASIC,
+    DTPI_OBJ_LAYOUT_SIMPLE__CONTIG,
+    DTPI_OBJ_LAYOUT_SIMPLE__VECTOR,
+    DTPI_OBJ_LAYOUT_SIMPLE__INDEXED,
+    DTPI_OBJ_LAYOUT_SIMPLE__BLOCK_INDEXED,
+    DTPI_OBJ_LAYOUT_SIMPLE__NUM
+ };
+
+/* complex derived datatype structures */
+enum {
+    DTPI_OBJ_LAYOUT_LARGE_BLK__VECTOR = DTPI_OBJ_LAYOUT_SIMPLE__NUM,
+    DTPI_OBJ_LAYOUT_LARGE_BLK__INDEXED,
+    DTPI_OBJ_LAYOUT_LARGE_BLK__BLOCK_INDEXED,
+    DTPI_OBJ_LAYOUT_LARGE_CNT__VECTOR,
+    DTPI_OBJ_LAYOUT_LARGE_CNT__INDEXED,
+    DTPI_OBJ_LAYOUT_LARGE_CNT__BLOCK_INDEXED,
+    DTPI_OBJ_LAYOUT_LARGE_BLK_STRD__VECTOR,
+    DTPI_OBJ_LAYOUT_LARGE_BLK_STRD__INDEXED,
+    DTPI_OBJ_LAYOUT_LARGE_BLK_STRD__BLOCK_INDEXED,
+    DTPI_OBJ_LAYOUT_LARGE_CNT_STRD__VECTOR,
+    DTPI_OBJ_LAYOUT_LARGE_CNT_STRD__INDEXED,
+    DTPI_OBJ_LAYOUT_LARGE_CNT_STRD__BLOCK_INDEXED,
+    DTPI_OBJ_LAYOUT__NUM
+};
+
+#define DTPI_VECTOR_NUM (5)
+#define DTPI_INDEXED_NUM DTPI_VECTOR_NUM
+#define DTPI_BLOCK_INDEXED_NUM DTPI_VECTOR_NUM
+#define DTPI_SIMPLE_NUM (3)
+#define DTPI_LARGE_BLOCK_NUM DTPI_SIMPLE_NUM
+#define DTPI_LARGE_COUNT_NUM DTPI_SIMPLE_NUM
+
+/* datatype position maps */
+extern int DTPI_Basic_map;
+extern int DTPI_Contig_map;
+extern int DTPI_Vector_map[DTPI_VECTOR_NUM];
+extern int DTPI_Indexed_map[DTPI_INDEXED_NUM];
+extern int DTPI_Block_indexed_map[DTPI_BLOCK_INDEXED_NUM];
+
+/* layout position maps */
+extern int DTPI_Simple_map[DTPI_SIMPLE_NUM];
+extern int DTPI_Large_block_map[DTPI_LARGE_BLOCK_NUM];
+extern int DTPI_Large_count_map[DTPI_LARGE_COUNT_NUM];
+
+typedef struct {
+    DTPI_obj_type_e obj_type;
+
+    MPI_Aint type_basic_size;
+    MPI_Aint type_extent;
+    MPI_Aint type_lb;
+    MPI_Aint type_ub;
+
+    union {
+        struct {
+            int num_blocks;
+        } vector;
+        struct {
+            int *displs;
+            int *num_blocks;
+        } indexed;
+        struct {
+            int *displs;
+            int num_blocks;
+        } block_indexed;
+        struct {
+            MPI_Aint *displs;
+        } structure;
+    } u;
+} DTPI_t;
+
+struct params {
+    int start;
+    int stride;
+    int count;
+    int obj_idx;
+    MPI_Aint type_count;
+    MPI_Aint type_blklen;
+    MPI_Aint type_stride;
+    MPI_Aint type_offset;
```

Comment:
It looks like the `type_` parameters are temporary ones that are internally set/updated. The others are user input, which are constant. Can you please make this difference clear ? E.g., you can put each group of parameters in different sub struct. This will be helpful for code reading.